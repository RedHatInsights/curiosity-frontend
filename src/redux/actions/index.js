import * as rhelActions from './rhelActions';
import * as userActions from './userActions';

const actions = {
  rhel: rhelActions,
  user: userActions
};

const reduxActions = { ...actions };

export { reduxActions as default, reduxActions, rhelActions, userActions };
