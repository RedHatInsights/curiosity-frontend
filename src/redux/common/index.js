import { reduxHelpers } from './reduxHelpers';

export { reduxHelpers as default, reduxHelpers };
