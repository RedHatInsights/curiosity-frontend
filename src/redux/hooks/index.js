import { reactReduxHooks } from './useReactRedux';

const storeHooks = {
  reactRedux: reactReduxHooks
};

export { storeHooks as default, storeHooks, reactReduxHooks };
