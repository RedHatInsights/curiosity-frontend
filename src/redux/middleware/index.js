import { createLogger } from 'redux-logger';
import promiseMiddleware from 'redux-promise-middleware';
import thunkMiddleware from 'redux-thunk';

const reduxMiddleware = [thunkMiddleware, promiseMiddleware];

if (process.env.NODE_ENV !== 'production' && process.env.REACT_APP_DEBUG_MIDDLEWARE === 'true') {
  reduxMiddleware.push(createLogger());
}

export { reduxMiddleware as default, reduxMiddleware, createLogger, promiseMiddleware, thunkMiddleware };
