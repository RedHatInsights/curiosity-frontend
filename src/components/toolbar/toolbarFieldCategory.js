import React from 'react';
import PropTypes from 'prop-types';
import { useProduct, useProductGraphConfig, useProductToolbarQuery } from '../productView/productViewContext';
import { reduxTypes, storeHooks } from '../../redux';
import { RHSM_API_QUERY_SET_TYPES } from '../../services/rhsm/rhsmConstants';
import { Select, SelectPosition } from '../form/select';
import { graphCardHelpers } from '../graphCard/graphCardHelpers';
import { translate } from '../i18n/i18n';

/**
 * A standalone Category select filter.
 *
 * @memberof Toolbar
 * @module ToolbarFieldCategory
 */

/**
 * Generate select field options from config
 *
 * @param {object} options
 * @param {Function} options.useProductGraphConfig
 * @returns {Function}
 */
const useToolbarFieldOptions = ({ useProductGraphConfig: useAliasProductGraphConfig = useProductGraphConfig } = {}) => {
  const { filters } = useAliasProductGraphConfig();
  const options = [];

  if (Array.isArray(filters)) {
    const updatedFilters = filters
      ?.map(({ metric, query }) => {
        const category = query?.[RHSM_API_QUERY_SET_TYPES.CATEGORY];

        if (category !== undefined) {
          return {
            title: translate('curiosity-toolbar.label', {
              context: ['category', (category === '' && 'none') || category]
            }),
            value: category,
            metaData: {
              metric,
              query
            },
            selected: false
          };
        }

        return undefined;
      })
      .filter(value => value !== undefined);

    if (updatedFilters?.length) {
      options.push(...updatedFilters);
    }
  }

  return options;
};

/**
 * On select update.
 *
 * @param {object} options
 * @param {Function} options.useDispatch
 * @param {Function} options.useProduct
 * @returns {Function}
 */
const useOnSelect = ({
  useDispatch: useAliasDispatch = storeHooks.reactRedux.useDispatch,
  useProduct: useAliasProduct = useProduct
} = {}) => {
  const { productId, viewId } = useAliasProduct();
  const dispatch = useAliasDispatch();

  return ({ value = null, selected = {} } = {}) => {
    let updatedGraphLegendValue = value;

    if (selected?.metaData?.metric) {
      updatedGraphLegendValue = graphCardHelpers.generateChartIds({
        metric: selected.metaData.metric,
        productId,
        query: { [RHSM_API_QUERY_SET_TYPES.CATEGORY]: value }
      });
    }

    dispatch([
      {
        type: reduxTypes.graph.SET_GRAPH_LEGEND,
        id: `${viewId}-inverted`,
        value: updatedGraphLegendValue
      },
      {
        type: reduxTypes.query.SET_QUERY_RESET_INVENTORY_LIST,
        viewId
      },
      {
        type: reduxTypes.query.SET_QUERY,
        viewId,
        filter: RHSM_API_QUERY_SET_TYPES.CATEGORY,
        value
      }
    ]);
  };
};

/**
 * Display a category field with generated options.
 *
 * @fires onSelect
 * @param {object} props
 * @param {boolean} props.isFilter
 * @param {string} props.position
 * @param {Function} props.t
 * @param {Function} props.useOnSelect
 * @param {Function} props.useProductToolbarQuery
 * @param {Function} props.useToolbarFieldOptions
 * @returns {React.ReactNode}
 */
const ToolbarFieldCategory = ({
  isFilter,
  position,
  t,
  useOnSelect: useAliasOnSelect,
  useProductToolbarQuery: useAliasProductToolbarQuery,
  useToolbarFieldOptions: useAliasToolbarFieldOptions
}) => {
  const { [RHSM_API_QUERY_SET_TYPES.CATEGORY]: updatedValue } = useAliasProductToolbarQuery();
  const onSelect = useAliasOnSelect();
  const options = useAliasToolbarFieldOptions();
  const updatedOptions = options.map(option => ({ ...option, selected: option.value === updatedValue }));

  return (
    <Select
      aria-label={t('curiosity-toolbar.placeholder', { context: [isFilter && 'filter', 'category'] })}
      onSelect={onSelect}
      options={updatedOptions}
      selectedOptions={updatedValue}
      placeholder={t('curiosity-toolbar.placeholder', { context: [isFilter && 'filter', 'category'] })}
      position={position}
      // variant={SelectVariant.checkbox}
      data-test="toolbarFieldCategory"
    />
  );
};

/**
 * Prop types.
 *
 * @type {{useOnSelect: Function, t: Function, useProductToolbarQuery: Function, isFilter: boolean,
 *     position: string, useToolbarFieldOptions: Function}}
 */
ToolbarFieldCategory.propTypes = {
  isFilter: PropTypes.bool,
  position: PropTypes.string,
  t: PropTypes.func,
  useOnSelect: PropTypes.func,
  useProductToolbarQuery: PropTypes.func,
  useToolbarFieldOptions: PropTypes.func
};

/**
 * Default props.
 *
 * @type {{useOnSelect: Function, t: translate, useProductToolbarQuery: Function, isFilter: boolean,
 *     position: string, useToolbarFieldOptions: Function}}
 */
ToolbarFieldCategory.defaultProps = {
  isFilter: false,
  position: SelectPosition.left,
  t: translate,
  useOnSelect,
  useProductToolbarQuery,
  useToolbarFieldOptions
};

export { ToolbarFieldCategory as default, ToolbarFieldCategory, useOnSelect, useToolbarFieldOptions };
