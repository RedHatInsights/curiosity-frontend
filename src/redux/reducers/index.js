import { combineReducers } from 'redux';
import { notifications } from '@redhat-cloud-services/frontend-components-notifications';
import graphReducer from './graphReducer';
import inventoryReducer from './inventoryReducer';
import viewReducer from './viewReducer';
import userReducer from './userReducer';

const reducers = {
  notifications,
  graph: graphReducer,
  inventory: inventoryReducer,
  user: userReducer,
  view: viewReducer
};

const reduxReducers = combineReducers(reducers);

export { reduxReducers as default, reduxReducers, graphReducer, inventoryReducer, userReducer, viewReducer };
