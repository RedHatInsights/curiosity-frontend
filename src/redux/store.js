import { createStore, applyMiddleware } from 'redux';
import { reduxMiddleware } from './middleware';
import { reduxReducers } from './reducers';

const store = createStore(reduxReducers, applyMiddleware(...reduxMiddleware));

export { store as default, store };
