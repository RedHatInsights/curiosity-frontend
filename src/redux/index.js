import { connect } from 'react-redux';
import { withRouter } from 'react-router-dom';
import { withTranslation } from 'react-i18next';
import { store } from './store';
import { reduxActions } from './actions';
import { reduxReducers } from './reducers';
import { reduxSelectors } from './selectors';
import { reduxTypes } from './types';
import { helpers } from '../common/helpers';

const connectTranslate = (mapStateToProps, mapDispatchToProps) => component =>
  connect(
    mapStateToProps,
    mapDispatchToProps
  )((!helpers.TEST_MODE && withTranslation()(component)) || component);

const connectRouterTranslate = (mapStateToProps, mapDispatchToProps) => component =>
  withRouter(connectTranslate(mapStateToProps, mapDispatchToProps)(component));

const connectRouter = (mapStateToProps, mapDispatchToProps) => component =>
  withRouter(
    connect(
      mapStateToProps,
      mapDispatchToProps
    )(component)
  );

export {
  connect,
  connectRouter,
  connectRouterTranslate,
  connectTranslate,
  reduxActions,
  reduxReducers,
  reduxSelectors,
  reduxTypes,
  store
};
