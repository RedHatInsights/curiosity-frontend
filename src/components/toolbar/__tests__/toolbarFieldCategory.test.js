import React from 'react';
import { ToolbarFieldCategory, useToolbarFieldOptions, useOnSelect } from '../toolbarFieldCategory';
import { store } from '../../../redux/store';
import {
  RHSM_API_QUERY_CATEGORY_TYPES as CATEGORY_TYPES,
  RHSM_API_QUERY_SET_TYPES
} from '../../../services/rhsm/rhsmConstants';

describe('ToolbarFieldCategory Component', () => {
  let mockDispatch;

  beforeEach(() => {
    mockDispatch = jest.spyOn(store, 'dispatch').mockImplementation((type, data) => ({ type, data }));
  });

  afterEach(() => {
    jest.clearAllMocks();
  });

  it('should render a basic component', async () => {
    const props = {
      useProductQuery: () => ({ [RHSM_API_QUERY_SET_TYPES.CATEGORY]: CATEGORY_TYPES.CLOUD })
    };

    const component = await shallowComponent(<ToolbarFieldCategory {...props} />);

    expect(component).toMatchSnapshot('basic');
  });

  it('should generate select options', async () => {
    const { result: toolbarFieldOptions } = await renderHook(() =>
      useToolbarFieldOptions({
        useProductGraphConfig: () => ({
          filters: [
            { metric: 'lorem', query: {} },
            { filters: [{ metric: 'ipsum', query: { [RHSM_API_QUERY_SET_TYPES.CATEGORY]: CATEGORY_TYPES.CLOUD } }] },
            { metric: 'dolor', query: { [RHSM_API_QUERY_SET_TYPES.CATEGORY]: CATEGORY_TYPES.VIRTUAL } },
            { metric: 'sit', query: { [RHSM_API_QUERY_SET_TYPES.CATEGORY]: CATEGORY_TYPES.HYPERVISOR } }
          ]
        })
      })
    );

    expect(toolbarFieldOptions).toMatchSnapshot('toolbarFieldOptions');
  });

  it('should handle updating through redux state with component', () => {
    const props = {
      useToolbarFieldOptions: () => [
        {
          metaData: {
            metric: 'dolor',
            query: {
              category: 'virtual'
            }
          },
          selected: false,
          title: 't(curiosity-toolbar.label_category, {"context":"virtual"})',
          value: 'virtual'
        },
        {
          metaData: {
            metric: 'sit',
            query: {
              category: 'hypervisor'
            }
          },
          selected: false,
          title: 't(curiosity-toolbar.label_category, {"context":"hypervisor"})',
          value: 'hypervisor'
        }
      ]
    };

    const component = renderComponent(<ToolbarFieldCategory {...props} />);
    const input = component.find('button');
    component.fireEvent.click(input);

    const inputMenuItem = component.find('button.pf-c-select__menu-item');
    component.fireEvent.click(inputMenuItem);

    expect(mockDispatch.mock.calls).toMatchSnapshot('dispatch, component');
  });

  it('should handle updating through redux state with hook', () => {
    const options = {
      useProduct: () => ({ productId: 'mock-product-id', viewId: 'mock-view-id' })
    };

    const onSelect = useOnSelect(options);
    onSelect({
      value: 'dolor sit',
      selected: {
        metaData: {
          metric: 'hello',
          query: {
            category: 'world'
          }
        }
      }
    });

    onSelect({
      value: 'lorem ipsum',
      selected: {}
    });

    expect(mockDispatch.mock.calls).toMatchSnapshot('dispatch, hook');
  });
});
