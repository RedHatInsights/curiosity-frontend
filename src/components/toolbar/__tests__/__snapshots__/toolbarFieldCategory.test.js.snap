// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ToolbarFieldCategory Component should generate select options: toolbarFieldOptions 1`] = `
[
  {
    "metaData": {
      "metric": "ipsum",
      "query": {
        "category": "cloud",
      },
    },
    "selected": false,
    "title": "t(curiosity-toolbar.label_category, {"context":"cloud"})",
    "value": "cloud",
  },
  {
    "metaData": {
      "metric": "dolor",
      "query": {
        "category": "virtual",
      },
    },
    "selected": false,
    "title": "t(curiosity-toolbar.label_category, {"context":"virtual"})",
    "value": "virtual",
  },
  {
    "metaData": {
      "metric": "sit",
      "query": {
        "category": "hypervisor",
      },
    },
    "selected": false,
    "title": "t(curiosity-toolbar.label_category, {"context":"hypervisor"})",
    "value": "hypervisor",
  },
]
`;

exports[`ToolbarFieldCategory Component should handle updating through redux state with component: dispatch, component 1`] = `
[
  [
    [
      {
        "id": "undefined-inverted",
        "type": "SET_GRAPH_LEGEND",
        "value": "dolor_virtual",
      },
      {
        "type": "SET_QUERY_RESET_INVENTORY_LIST",
        "viewId": undefined,
      },
      {
        "filter": "category",
        "type": "SET_QUERY",
        "value": "virtual",
        "viewId": undefined,
      },
    ],
  ],
]
`;

exports[`ToolbarFieldCategory Component should handle updating through redux state with hook: dispatch, hook 1`] = `
[
  [
    [
      {
        "id": "mock-view-id-inverted",
        "type": "SET_GRAPH_LEGEND",
        "value": "hello_dolor sit_mock-product-id",
      },
      {
        "type": "SET_QUERY_RESET_INVENTORY_LIST",
        "viewId": "mock-view-id",
      },
      {
        "filter": "category",
        "type": "SET_QUERY",
        "value": "dolor sit",
        "viewId": "mock-view-id",
      },
    ],
  ],
  [
    [
      {
        "id": "mock-view-id-inverted",
        "type": "SET_GRAPH_LEGEND",
        "value": "lorem ipsum",
      },
      {
        "type": "SET_QUERY_RESET_INVENTORY_LIST",
        "viewId": "mock-view-id",
      },
      {
        "filter": "category",
        "type": "SET_QUERY",
        "value": "lorem ipsum",
        "viewId": "mock-view-id",
      },
    ],
  ],
]
`;

exports[`ToolbarFieldCategory Component should render a basic component: basic 1`] = `
<Select
  aria-label="t(curiosity-toolbar.placeholder, {"context":"category"})"
  ariaLabel="Select option"
  buttonContent={null}
  className=""
  data-test="toolbarFieldCategory"
  direction="down"
  id="generatedid-"
  isDisabled={false}
  isDropdownButton={false}
  isFlipEnabled={false}
  isInline={true}
  isToggleText={true}
  maxHeight={null}
  name={null}
  onSelect={[Function]}
  onSplitButton={null}
  options={[]}
  placeholder="t(curiosity-toolbar.placeholder, {"context":"category"})"
  position="left"
  selectedOptions={null}
  splitButtonAllowDualButtonToggle={true}
  splitButtonVariant={null}
  toggleIcon={null}
  variant="single"
/>
`;
